### PR TITLE
unpack ICS008 wasm client state at ConnOpenTry and ConnOpenAck

### LINKED ISSUE #1237: support ICS008 wrapper type for host client and consensus state
## Feature Summary

Now that ibc-go comes with wasm-08, client state of an ibc-rs chain maybe stored as wasm wrapped format in wasm-08 enabled chain.

So when `MsgConnectionOpenTry` or `MsgConnectionOpenAck` is received at ibc-rs, it should expect two cases:
1. normal client state. e.g., the chain is using vanilla tendermint engine, so the counterparty ibc-go chain uses ibc-go implementation of ics07.
2. wasm wrapped client state. e.g., the chain is using a custom consensus, so the counterparty ibc-go chain uses wasm client via wasm-08

## Proposal

We need to introduce an enum:
```rs
pub enum HostClientState<T> {
    Native {
        state: T,
    },
    Wasm {
        state: T,
        checksum: Vec<u8>,
        latest_height: Height,
    },
}
```

Then, use `HosClientState<Ctx::HostClientState>` in the following places:

https://github.com/cosmos/ibc-rs/blob/2378cd4ba45094b8ed856ff7dba5f1d0882f59ae/ibc-core/ics03-connection/src/handler/conn_open_ack.rs#L52

https://github.com/cosmos/ibc-rs/blob/2378cd4ba45094b8ed856ff7dba5f1d0882f59ae/ibc-core/ics03-connection/src/handler/conn_open_try.rs#L42

### PR DESCRIPTION
**Pull request by rnbguy**
Closes: #1237

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-core/ics03-connection/src/handler/mod.rs [hunk 1]
```diff
@@ -1,4 +1,50 @@
...
+use ibc_core_handler_types::error::ContextError;
+use ibc_core_host::types::identifiers::ClientId;
+use ibc_primitives::proto::Any;
+
 pub mod conn_open_ack;
 pub mod conn_open_confirm;
 pub mod conn_open_init;
 pub mod conn_open_try;
+
+pub(crate) fn unpack_host_client_state<CS>(
```
Comment (seanchen1991) [inline ibc-core/ics03-connection/src/handler/mod.rs:16]: Could we add a doc string for this method?

File: ibc-core/ics24-host/types/src/identifiers/client_id.rs [hunk 2]
```diff
@@ -66,6 +66,16 @@ impl ClientId {
     pub fn as_bytes(&self) -> &[u8] {
         self.0.as_bytes()
     }
+
+    /// Check if the client identifier is for 08-wasm light client.
+    pub fn is_wasm_client_id(&self) -> bool {
```
Comment (seanchen1991) [inline ibc-core/ics24-host/types/src/identifiers/client_id.rs:71]: Perhaps define a string constant for the `"08-wasm-"` prefix?

**Review by seanchen1991 — APPROVED** [review]

File: ibc-core/ics03-connection/src/handler/mod.rs (hunk 1 above)
Comment (rnbguy) [reply ibc-core/ics03-connection/src/handler/mod.rs:16]: ebd2327

File: ibc-core/ics24-host/types/src/identifiers/client_id.rs (hunk 2 above)
Comment (rnbguy) [reply ibc-core/ics24-host/types/src/identifiers/client_id.rs:71]: 43b88a6

**Pull request merged**